File: Algorithms/python/bubble_sort.py
"""
========================================================================
@file bubblesort.py

: zach wolpe
: zach.wolpe@medibio.com.au
: 17 July 2023
========================================================================
"""


def bubble_sort(array):
    """
    :param array: list of numbers to be sorted
    :return: sorted list
    """
    
    for i in range(len(array)-1, 0, -1):
        for j in range(i):
            if array[j] > array[j+1]:
                temp        = array[j]
                array[j]    = array[j+1]
                array[j+1]  = temp
    return array

if __name__ == '__main__':
    
    array = [1, 222, 233, 4, 5, 6, 97, 8, 9]
    array = bubble_sort(array)
    [print(i, end=' -> ') for i in array]
    print()

File: Algorithms/cpp/bubble_sort.cpp
/**
========================================================================
@file bubblesort.cpp

: zach wolpe
: zach.wolpe@medibio.com.au
: 14 July 2023
========================================================================
*/

#include <iostream>
using namespace std;

void bubbleSort(int array[], int size) {
    // passing array only passes the address of the first item in the array, not the size. 
    // we do not return a new array, but update an existing one.
    for (int i=size-1; i>0; i--) {
        for (int j=0; j<i; j++) {
            if (array[j] > array[j+1]) {
                int temp    = array[j];
                array[j]    = array[j+1];
                array[j+1]  = temp;
            } 
        }
    }
}

int main() {
    int myArray[] = {6,4,2,5,3,4,2};
    int size = sizeof(myArray) / sizeof(myArray[0]); // number of bytes in array
    bubbleSort(myArray, size);

    for (auto value : myArray) {
        cout << value << " ";
    }
    return 0;
};